### PR TITLE
Ordering

### PR DESCRIPTION
**Pull request by Alfredbis29**
This PR adds the ability for users to place orders for food items in the app.

-  Add Order model to store details of each order 
- Add `OrdersController` with actions for creating and displaying orders.
- Created order action in foodsController to render form for selecting and submitting orders.
- Add routes for creating and displaying orders.
- Add functionality to calculate and display the total cost of an order.

### CODE REVIEW

File: app/controllers/orders_controller.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+   @order = Order.all
+  end
+
+  def new
+    @order = Order.new
+    @foods = Food.all
+  end
+
+  def create
+    @order = Order.new(order_params)
```
Comment (Tresor11) [inline app/controllers/orders_controller.rb:18]: this can be `@order = current_user.orders.new(order_params)`

File: app/models/food.rb [hunk 2]
```diff
@@ -1,3 +1,5 @@
 class Food < ApplicationRecord
   belongs_to :user
+has_many :orders
```
Comment (Tresor11) [inline app/models/food.rb]: Can you fix the indentation here?

Comment (Tresor11) [inline app/models/food.rb]: The food should belong to an order

File: app/models/order.rb [hunk 3]
```diff
@@ -0,0 +1,4 @@
+class Order < ApplicationRecord
+  belongs_to :user
+  belongs_to :food
```
Comment (Tresor11) [inline app/models/order.rb:3]: It should have many foods

File: test/factories/orders.rb [hunk 4]
```diff
@@ -0,0 +1,6 @@
+FactoryBot.define do
+  factory :order do
+    user { nil }
```
Comment (Tresor11) [inline test/factories/orders.rb]: I think we need a user here same this for the food

File: config/routes.rb [hunk 5]
```diff
@@ -1,7 +1,9 @@
 Rails.application.routes.draw do
   # get 'dashboards/index'
   root 'dashboards#index'
+  resources :orders
 
   devise_for :users
   resources :foods, only: %i[index show new create destroy]
```
Comment (Tresor11) [inline config/routes.rb:7]: Why did you add the `only` option here?

File: app/models/user.rb [hunk 6]
```diff
@@ -4,6 +4,9 @@ class User < ApplicationRecord
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :validatable, :confirmable
 
+  has_many :orders
+  has_many :foods, through: :orders
```
Comment (Tresor11) [inline app/models/user.rb]: Nice!

File: config/routes.rb (hunk 5 above)
Comment (Alfredbis29) [reply config/routes.rb:7]: @Tresor11 sorry I don't really get

File: app/models/food.rb (hunk 2 above)
Comment (vic778) [reply app/models/food.rb]: in fact, I think the order should belong to the food and the user. because we need the food_id and user_id to create the order.

File: app/views/orders/index.html.erb [hunk 7]
```diff
@@ -2,17 +2,27 @@
...
+   <%= form.label :image, style: "display: block" %>
+   <%= form.file_field :image %>
+  </div><br/>
 
-  <%= form.label :price %>
-  <%= form.number_field :price %>
+    <div>
+      <%= form.label :price, style: "display: block" %>
+      <%= form.text_field :price %>
+    </div><br/>
```
Comment (vic778) [inline app/views/orders/index.html.erb]: We don't need these fields. instead, we could have the quantity itself because the food_id contains all this information as well.

File: app/models/user.rb [hunk 8]
```diff
@@ -3,9 +3,8 @@ class User < ApplicationRecord
   # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :validatable, :confirmable
-
-  has_many :orders
-  has_many :foods, through: :orders
+         has_many :orders, class_name: "Order"
```
Comment (Tresor11) [inline app/models/user.rb:6]: Why are we adding the class name here?

File: db/migrate/20230117161036_create_orders.rb [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+    create_table :orders do |t|
+      t.integer :quantity
+      t.string :ref_code
+      t.float :total_price
+      t.string :status
+
+      t.timestamps
+    end
+  end
+end
```
Comment (vic778) [inline db/migrate/20230117161036_create_orders.rb:12]: well done. it can be better to associate it with the user and food. so we will be able to see who performed the order and which food was ordered

File: app/models/order.rb [hunk 10]
```diff
@@ -0,0 +1,2 @@
+class Order < ApplicationRecord
+end
```
Comment (vic778) [inline app/models/order.rb]: Don't forget to add the belongs_to

File: db/migrate/20230117161036_create_orders.rb [hunk 11]
```diff
@@ -5,8 +5,13 @@ def change
       t.string :ref_code
       t.float :total_price
       t.string :status
+      t.integer :user_id
+      t.integer :food_id
+
```
Comment (vic778) [inline db/migrate/20230117161036_create_orders.rb]: Delete these lines. we don't need them

File: db/migrate/20230117161036_create_orders.rb [hunk 12]
```diff
@@ -5,8 +5,13 @@ def change
...
 
       t.timestamps
     end
+    add_index :orders, :user_id
+    add_index :orders, :food_id
```
Comment (vic778) [inline db/migrate/20230117161036_create_orders.rb]: And this as well

File: db/migrate/20230117161324_add_user_id_to_orders.rb [hunk 13]
```diff
@@ -0,0 +1,5 @@
+class AddUserIdToOrders < ActiveRecord::Migration[7.0]
+  def change
+    add_column :orders, :user_id, :integer
+  end
```
Comment (vic778) [inline db/migrate/20230117161324_add_user_id_to_orders.rb]: I think the better could be to add a reference instead of a column like 
`add_reference :orders, :user, foreign_key: true`

File: db/migrate/20230117161800_add_food_id_to_orders.rb [hunk 14]
```diff
@@ -0,0 +1,5 @@
+class AddFoodIdToOrders < ActiveRecord::Migration[7.0]
+  def change
+    add_column :orders, :food_id, :integer
+  end
```
Comment (vic778) [inline db/migrate/20230117161800_add_food_id_to_orders.rb]: the same here.

File: app/views/menu_items/menu_items.html.erb [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+  <%= form.label :menu_items %>
+  <% MenuItem.all.each do |menu_item| %>
+    <div>
+      <%= menu_item.name %>
+      <%= form.number_field "menu_items[#{menu_item.id}]", min: 0 %>
+    </div>
+  <% end %>
+
+  <%= form.submit %>
+<% end %>
```
Comment (vic778) [inline app/views/menu_items/menu_items.html.erb]: Can you explain to me what does this code ?
